packages/rollup-full-node/src/app/message-submitter.ts

```diff
@@ -0,0 +1,289 @@
+        log.info(`No messages queued for submission.`)
+      } else {
+        log.debug(
+          `Next message queued but not submitted because message ${this.lastSubmitted} was submitted but not yet confirmed.`
+        )
+      }
+
+      return
+    }
+
+    const message: L2ToL1Message = this.messageQueue[0]
+
+    log.info(
+      `Submitting message number ${message.nonce}: ${JSON.stringify(message)}.`
+    )
+
+    let receipt
+    try {
+      receipt = await this.messageReceiverContract.enqueueL2ToL1Message(
+        DefaultL2ToL1MessageSubmitter.serializeRollupMessageForSubmission(
+          message
+        )
+      )
+      // TODO: do something with receipt?
+    } catch (e) {
+      logError(
+        log,
+        `Error submitting rollup message: ${JSON.stringify(message)}`,
+        e
+      )
+      throw e
+    }
+
+    this.lastSubmitted = message.nonce
+    await this.db.put(
+      DefaultL2ToL1MessageSubmitter.LAST_SUBMITTED_KEY,
+      this.lastSubmitted.toBuffer()
+    )
+
+    this.messageReceiverContract.provider
+      .waitForTransaction(receipt)
+      .then((txReceipt: TransactionReceipt) => {
+        log.debug(
+          `L2 -> L1 Message with nonce ${message.nonce.toString(
+            'hex'
+          )} was confirmed on L1!`
+        )
+      })
+      .catch((error) => {
+        logError(log, 'Error submitting L2 -> L1 message transaction', error)
+      })
+      .finally(async () => {
+        this.lastConfirmed = message.nonce
+        await this.db.put(
+          DefaultL2ToL1MessageSubmitter.LAST_CONFIRMED_KEY,
+          this.lastConfirmed.toBuffer()
+        )
+
+        // If we failed after submission but before storing submitted, update lastSubmitted
+        if (this.lastSubmitted.lt(this.lastConfirmed)) {
+          this.lastSubmitted = message.nonce
+          await this.db.put(
+            DefaultL2ToL1MessageSubmitter.LAST_SUBMITTED_KEY,
+            this.lastSubmitted.toBuffer()
+          )
+        }
+        // purposefully don't await
+        this.trySubmitNextMessage()
+      })
+  }
+
+  public static serializeRollupMessageForSubmission(
+    rollupMessage: L2ToL1Message
+  ): string {
+    return abiEncodeL2ToL1Message(rollupMessage)
+  }
+
+  public static serializeRollupMessageForStorage(
+    rollupMessage: L2ToL1Message
+  ): Buffer {
+    const rollupMessageString: string = abiEncodeL2ToL1Message(rollupMessage)
+    return Buffer.from(rollupMessageString)
+  }
+
+  public static deserializeRollupMessageFromStorage(
+    rollupMessageBuffer: Buffer
+  ): L2ToL1Message {
+    const serializedRollupMessage: string = rollupMessageBuffer.toString()
+    return abiDecodeL2ToL1Message(serializedRollupMessage)
+  }
+
+  public static getMessageKey(messageNumber: BigNumber): Buffer {
+    return Buffer.from(`MESSAGE_${add0x(messageNumber.toString('hex'))}`)
+  }
+
+  /***********
+   * GETTERS *
+   ***********/
+  public getLastSubmitted(): BigNumber {
```

Comment:
Any interest in using [TypeScript Accessors](https://www.typescriptlang.org/docs/handbook/classes.html#accessors)?
```suggestion
  public get lastSubmitted(): BigNumber {
```
I'm also open to the idea of dropping getters completely and just accessing state directly 😬 That's probably a little to `YOLO` though.